### ISSUE TITLE
Site Editor Feature Request: See which templates use a given template part 

### ISSUE THREAD

**Issue by jp-imagines** [WordPress/gutenberg#42482]
## What problem does this address?

On FSE/block themes, it's currently difficult to quickly determine which template parts are actively in use, or which templates use any given template part. 

For example, if I want to update my site's header across all templates, and I have multiple custom header template parts available, I would need to either:

* Manually check every _template_ and update the "header" template part being used in each; or
* Update every "header" _template part_, some of which may not be in use.

## What is your proposed solution?

Within the Site Editor, possibly on the list of _Template Parts_, it would be great to see a full list of templates that use any given template part (if any). (Optionally, a similar list of all template parts being used on each template on the _Templates_ list would provide similar helpful information.)

This would also make it much clearer that updating _Template Part A_ would affect _Templates X and Y_.

Quick mock-up:
![template part mockup](https://user-images.githubusercontent.com/5630717/177846463-08c174f4-7718-4f09-b055-f1448a75885d.png)

**Comment by talldan** [comment]
This would also be nice in a warning message when deleting a template part that's in use.

Code-wise, my understanding is that it would involve parsing the block content of every template and template part to find out which template parts are within them.

It'd be challenging and probably quite expensive to do.

**Comment by Andrew-Starr** [comment]
I have a local plugin I created that does this. It is very helpful to see at a glance what template part is in use in which template.

The code is a bit rough but it basically parses the content of every template, then "flattens" the parsed content using the core `_flatten_blocks` function, and finally checks all the returned blocks to see if any have a `blockName` of `core/template-part` and returns an array of the found 'slug' 'attribute' for each part.

Repeat the above for template parts within template parts too.

I can't speak to how expensive it is but have not seen any noticable slow down doing this on my own local test install.

**Comment by Andrew-Starr** [comment]
I should also say that this is done in a separate admin page outside of the site editor for my own usage, and so if something like this was included in the site editor I do not know how resource intensive this would be and whether or not the site editor has the scalability for this.

**Comment by jameskoster** [comment]
I agree this would be a useful enhancement. With all the work we're doing to better surface template part creation via patterns, it is easy for a proliferation of un-used template parts to occur.

**Comment by jameskoster** [comment]
One issue we might run into here is the significant visual noise when there are many templates. I do acknowledge there probably is some value in listing the templates, potentially even as links to go and edit them. Maybe we could do something like:

<img width="1212" alt="Screenshot 2022-07-29 at 12 31 21" src="https://user-images.githubusercontent.com/846565/181750137-9e782203-4e67-479f-8428-c0c7e70d62de.png">


Side note: It might be nice to do something similar for custom templates in the main template list:

<img width="1212" alt="Screenshot 2022-07-29 at 12 31 30" src="https://user-images.githubusercontent.com/846565/181750170-59cfc3a6-357c-47f4-ab6b-d8233cf9f4e2.png">

**Comment by annezazu** [comment]
This was noted as part of the [FSE Outreach Program's Site Editor Sneak Peek exploration](https://make.wordpress.org/test/2022/11/13/fse-program-exploration-site-editor-sneak-peek/#comment-2765): 

>When editing and saving a template part, it would be nice to know which templates each template part is included in. It wasn’t immediately clear that this would update any existing template that it is already used in (vs. making it available to be used for new templates)

**Comment by carolinan** [comment]
How would this menu look when there is more than one item? For example the delete template action.

**Comment by tyxla** [comment]
Closing this as duplicate of #20476.
